File: share/example/c++/example_freedrive.cpp
#include "aubo_sdk/rpc.h"
#ifdef WIN32
#include <windows.h>
#endif

using namespace arcs::aubo_sdk;

#define LOCAL_IP "127.0.0.1"

int main(int argc, char **argv)
{
#ifdef WIN32
    // Set Windows console output code page to UTF-8
    SetConsoleOutputCP(CP_UTF8);
#endif
    // RPC client object
    auto rpc_cli = std::make_shared<arcs::aubo_sdk::RpcClient>();
    // API call: set RPC timeout
    rpc_cli->setRequestTimeout(1000);
    // API call: connect to RPC service
    rpc_cli->connect(LOCAL_IP, 30004);
    // API call: login
    rpc_cli->login("aubo", "123456");
    // API call: get robot name
    auto robot_name = rpc_cli->getRobotNames().front();
    auto impl = rpc_cli->getRobotInterface(robot_name);

    // API call: enable freedrive mode
    impl->getRobotManage()->freedrive(true);

    // Wait to enter freedrive mode
    int i = 0;
    while (!impl->getRobotManage()->isFreedriveEnabled()) {
        if (i++ > 5) {
            std::cerr << "Failed to enable freedrive mode" << std::endl;
            return -1;
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(5));
    }
    std::cout << "Freedrive mode enabled successfully" << std::endl;

    std::this_thread::sleep_for(std::chrono::seconds(25));

    // API call: exit freedrive mode
    impl->getRobotManage()->freedrive(false);

    // Wait to exit freedrive mode
    while (impl->getRobotManage()->isFreedriveEnabled()) {
        if (i++ > 5) {
            std::cerr << "Failed to disable freedrive mode" << std::endl;
            return -1;
        }
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
    std::cout << "Freedrive mode disabled successfully" << std::endl;

    // API call: logout
    rpc_cli->logout();
    // API call: disconnect
    rpc_cli->disconnect();

    return 0;
}

File: share/example/python/example_freedrive.py
#! /usr/bin/env python
# coding=utf-8

"""
Freedrive Teaching

Steps:
Step 1: Connect to the RPC server and log in
Step 2: Enter freedrive teaching mode, exit the mode after 25 seconds
"""
import pyaubo_sdk
import time

robot_ip = "127.0.0.1"  # Server IP address
robot_port = 30004  # Port number
M_PI = 3.14159265358979323846
robot_rpc_client = pyaubo_sdk.RpcClient()


def exampleFreeDrive(cli):
    robot_name = cli.getRobotNames()[0]
    cli.getRuntimeMachine().start()
    # API call: Initiate robot freedrive request
    cli.getRobotInterface(robot_name).getRobotManage().freedrive(True)
    time.sleep(25)
    # API call: End robot freedrive request
    cli.getRobotInterface(robot_name).getRobotManage().freedrive(False)
    pass


if __name__ == '__main__':
    robot_rpc_client.setRequestTimeout(1000)
    robot_rpc_client.connect(robot_ip, robot_port)  # API call: Connect to RPC service
    if robot_rpc_client.hasConnected():
        print("Robot rcp_client connected successfully!")
        robot_rpc_client.login("aubo", "123456")  # API call: Robot login
        if robot_rpc_client.hasLogined():
            print("Robot rcp_client logined successfully!")
            exampleFreeDrive(robot_rpc_client)  # Enter freedrive teaching mode, exit after 25 seconds
